File: src/Y2018/facebook/Q031_Next_Permutation.java
package Y2018.facebook;

import java.util.Arrays;

public class Q031_Next_Permutation {

    public void nextPermutation(int[] nums) {
        if (nums == null || nums.length <= 1) return;
        int i = nums.length - 2;
        while (i >= 0 && nums[i] >= nums[i + 1]) i--;
        if (i >= 0) {
            int j = nums.length - 1;
            while (nums[j] <= nums[i]) j--;
            swap(nums, i, j);
        }
        // reverse(nums, i+1, nums.length-1);
        Arrays.sort(nums, i + 1, nums.length);
    }

    private void reverse(int[] nums, int l, int r) {
        while (l < r) {
            swap(nums, l++, r--);
        }
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

}

File: leetcode-go/Q031-Next-Permutation/Q031-Next-Permutation.go
package Q031_Next_Permutation

import (
	"sort"
)

func nextPermutation(nums []int) {
	if len(nums) < 2 {
		return
	}

	first := -1
	for i := len(nums) - 1; i >= 1; i-- {
		if nums[i-1] < nums[i] {
			first = i - 1
			break
		}
	}
	if first == -1 {
		sort.Ints(nums)
		return
	}

	second := -1
	for i := len(nums) - 1; i > first; i-- {
		if nums[i] > nums[first] {
			second = i
			break
		}
	}

	nums[first], nums[second] = nums[second], nums[first]
	sort.Ints(nums[first+1:])
}
